### PR TITLE
[Feat](bangc-ops): add new op nms_rotated

### PR DESCRIPTION
**Pull request by liuyuan1-v**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

新增算子 nms_rotated

## 2. Modification

影响范围/算子：nms_rotated
影响版本/分支：master, r0.4

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

算子采用静态阈值标准: diff3 = 0

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |        MLU290<br>MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1       |
|        3       |Layouts                               |        ARRAY      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                支持                      |
|        6       |Data type(half/float)                 |           float           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [ ] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |    1. box和score对应维度不一致：  [MLUOP] [Error]:[mluOpNmsRotated] Check failed: boxes_desc->dims[0] == scores_desc->dims[0].   <br> 2. box的低维不等于5或6： [MLUOP] [Error]:[mluOpNmsRotated] Check failed: The Boxes' last dimenstion should be 5 or 6. Now is 7. <br>3. 数据类型不为float： [MLUOP] [Error]:[mluOpNmsRotated] Check failed: boxes_desc->dtype == MLUOP_DTYPE_FLOAT. <br> 4. 数据类型不一致： [MLUOP] [Error]:[mluOpNmsRotated] Check failed: boxes_desc->dtype == scores_desc->dtype. <br> 5. 输出类型不为int32：[MLUOP] [Error]:[mluOpNmsRotated] Check failed: output_desc->dtype == MLUOP_DTYPE_INT32. <br> 6.输入box的维度不为2： [MLUOP] [Error]:[mluOpNmsRotated] Check failed: boxes_desc->dim == 2.  |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test   | half/float/int8              |  [-------------]  Global test environment tear-down  [ SUMMARY  ] Total 1000 cases of 1 op(s). ALL PASSED. <br> [=========] 1000 test cases from 1 test suite ran. (717165 ms total)  [  PASSED  ] 1000 test cases.    |     |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |    支持0元素，返回success     |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |     不支持box, score 输入 inf/nan     |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU290

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
| nms_rotated  |  6.36e5  |  77.19  |  2.79e-9  |  3.59e-6  | 1872  | float  | [78,5]    |
| nms_rotated  |  2.72e5  |  83.99  |  2.84e-9  |  3.01e-6  |  816  |  float  |  [34,5]   |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
| nms_rotated | 3.09e5  |  68.24  |  3.45e-9  |  6.56e-6  |  1872  |  float  |  [78,5]   |
| nms_rotated | 1.13e5  |  66.19  |  4.13e-9  |  6.48e-6  |  816    |  float  |  [34, 5]  |

### 3.4 Summary Analysis

内存泄漏测试：
./build.sh --asan 编译成功并且运行正常

覆盖率测试：
Filename [Sort by      Line Coverage [Sort_by_line_coverage] Functions [Sort_by
name]                                                        function_coverage]
nms_rotated_union1.mlu [96.7%][96.7%] 96.7 %384 / 397       100.0 %3 / 3
nms_utils.h            [98.3%][98.3%] 98.3 %686 / 698       100.0 %9 / 9

### CODE REVIEW

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+### 1.1 算子需求分析
+  
+| 算子功能简介               | 简要填写算子功能，详细描述在 1.2 中进行说明                           |
+| ------------------------ | -------------------------------------------------------------  |
+| 需求来源                  | mmcv                                                           |
+| 应用网络                  | fcos3d/pointpillar                                             |
+| 输入数据类型               | float                                                          |
+| 输入 Shape                | dets:[N,5]; scores:[N]; multi_label:int; iou_threshold:float;  |
+| 输出数据类型               | int64                                                          |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:41]: 这些都跟原始需求对齐了嘛。

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    5. 计算有效的交叠面积 intersection 后，计算结果`IOU`。
+3. 将`IOU`大于`iou_threshold`的box移除，回到step1直到对所有的`scores`完成遍历。
+
+### 1.3 算子输入输出参数要求
+
+| 参数           | 语义                       | 类型（输入/输出） | 支持类型                 | 物理布局        | 规模限制 |
+| -----------   | -------------------------- | -------------- | ----------------------- | ------------- | -------- |
+| handle        | 操作句柄                    | 输入            | mluOpHandle_t           | /             | /        |
+| dets_desc     | 输入数据，box 的描述符        | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| dets          | 输入数据，box 的坐标          | 输入            | float/half\*            | [N, 5]        | /        |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:69]: dets都改成boxes？dets 是个缩写？看不懂是啥意思

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+| scores_desc   | 输入数据，scores 的描述符     | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| scores        | 输入数据，scores 的大小       | 输入            | float/half\*            | [N]           | /        |
+| workspace     | 输入数据，GDRAM上面的辅助空间  | 输入            | void *\*                | /             | /        |
+| workspace_size| 输入数据，辅助空间的大小       | 输入            | size_t\*                | /             | /        |
+| iou_threshold | 输入数据，IOU 的阈值          | 输入            | float                   | scalar        | /        |
+| inds_desc     | 输入数据，输出 index 的描述符  | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| inds          | 输出数据，输出 index 的数据    | 输出            | int32\*                 | [N]           | /        |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:76]: inds 也是缩写，参考mluOpPolyNms接口参数命名，建议改成output

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+| result_num    | 输出数据，输出 box 的个数      | 输出            | int32 *\*               | /             | /        |
+
+### 1.4 算子限制
+
+| 限制类型     | 详细说明                                                                         |
+| ----------- | ------------------------------------------------------------------------------ |
+| 输入限制     | 输入参数`dets`，`scores`的shape 必须满足要求: dets[N, 5]和 scores:[N]               |  
+| 输出限制     | 输出参数`out_inds`的shape 必须满足: [N]                                            |
+| 输入限制     | 输入 `iou_threshold` 仅支持输入float                                              |
+| 输入限制     | 输入 `multi_label` 仅支持 int32, 该参数目前对结果不产生影响                           |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:86]: multi_label 是接口里哪个参数？作用是干啥，需要预留嘛？

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+| dets_desc     | 输入数据，box 的描述符        | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| dets          | 输入数据，box 的坐标          | 输入            | float/half\*            | [N, 5]        | /        |
+| scores_desc   | 输入数据，scores 的描述符     | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| scores        | 输入数据，scores 的大小       | 输入            | float/half\*            | [N]           | /        |
+| workspace     | 输入数据，GDRAM上面的辅助空间  | 输入            | void *\*                | /             | /        |
+| workspace_size| 输入数据，辅助空间的大小       | 输入            | size_t\*                | /             | /        |
+| iou_threshold | 输入数据，IOU 的阈值          | 输入            | float                   | scalar        | /        |
+| inds_desc     | 输入数据，输出 index 的描述符  | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| inds          | 输出数据，输出 index 的数据    | 输出            | int32\*                 | [N]           | /        |
+| result_num    | 输出数据，输出 box 的个数      | 输出            | int32 *\*               | /             | /        |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:77]: 这个参数列表好像不全

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 4 above)
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:86]: multi_label 不同主要有啥区别？

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+| 输出数据类型               | int64                                                          |
+| 输出 Shape                | inds:[N], result_num: int32, 表示输出box的个数                   |
+| 是否需要支持原位            | 否                                                              |
+| 是否需要支持 stride 机制    | 否                                                              |
+| 是否需要支持广播            | 否                                                              |
+| 0 元素检查是否直接返回       | 是                                                              |
+
+### 1.2 算子功能和应用场景描述
+
+NmsRotated 算子有 2 个输入 Tensor，分别为 `dets`[N,5], `scores`[N]，`dets`输入低维度的 5 个数字包括的信息为： `x_ctr`, `y_ctr`, `height`, `width`, `radian`.
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:50]: 参考https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/csrc/common/cuda/nms_rotated_cuda.cuh#L42， `x_ctr`, `y_ctr`, `height`, `width`, `radian` -》 (x_center, y_center, width, height, angle_degrees)

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+| stride 限制 | 不支持 stride 机制                                                                |
+| 广播限制     | 不支持广播                                                                        |
+
+### 1.5 验收标准
+
+#### 1.5.1 精度验收标准
+
+- 该算子输出`inds`为所选择的box的索引数据。`inds`是 int32_t 数据类型。因此，该算子采用静态阈值，阈值标准：diff3 = 0.
+
+- 注意：MLU200 系列精度需要限定数值范围和规模大小，避免计算IOU时出现大规模随机错误。
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:98]: 对于真值的限制说明放到1.4小节：（1）inf/nan 也说明各平台的行为；（2）score相同时，选择的框可能跟cpu、gpu不同 （3）其他的限制

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
+# nms rotated 算子开发设计方案
+
+- #### 文档基本信息
+
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:4]: 添加 /user_guide/9_operators/index.rst 处内容

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
                    const mluOpTensorDescriptor_t ious_desc,
                    void *ious);
 
+// Group:NmsRotated
+/*!
+ * @brief Computes the index of nms with intersection-over-union
+ * (Jaccard index, IOU) of rotated bounding-boxes.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4376]: Computes the index of nms with IOU of rotated bounding boxes.

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and
+ * queues in the nms_rotated operation. For detailed information, see
+ * ::mluOpHandle_t.
+ * @param[in] iou_threshold
+ * The threshold of iou.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4383]: iou > IOU

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] boxes_desc
+ * The descriptor of the input tensor \b boxes (rotated bounding-box).
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4385]: (rotated bounding boxes)

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the input tensor \b dets.
+ * It has shape (n, 5) or (n, 6), indicating (x, y, w, h, theta) or
+ * (x, y, w, h, theta, label) for each row. Note that theta is in radian.
+ * @param[in] scores_desc
+ * The descriptor of the input tensor \b scores (rotated bounding-box).
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4392]: (rotated bounding boxes)

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * Handle to an MLUOP context that is used to manage MLU devices and
+ * queues in the nms_rotated operation. For detailed information, see
+ * ::mluOpHandle_t.
+ * @param[in] iou_threshold
+ * The threshold of iou.
+ * @param[in] boxes_desc
+ * The descriptor of the input tensor \b boxes (rotated bounding-box).
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the input tensor \b dets.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4388]: dets这个没有问题吧？

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] scores_desc
+ * The descriptor of the input tensor \b scores (rotated bounding-box).
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] scores
+ * Pointer to the MLU memory that stores the input tensor \b scores.
+ * It has shape (n), indicating score of each box in \b boxes.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the Nms operation.
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that need to be used in the Nms operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4400]: needs

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] output_desc
+ * The descriptor of the output tensor. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] output
+ * Pointer to the MLU memory that stores the output tensor, which indicates
+ * the index of each output boxes.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4406]: boxes > box

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 1 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已对齐，按照int32类型输出

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 2 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已修改

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 3 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已修改

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 5 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已添加iou_threshold,已删除 multi_label

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 4 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: multi_label 已删除，通过box的描述符获取信息

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 6 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已修改

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 7 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:99]: 已添加限制说明

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 8 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:4]: 已补充

File: bangc-ops/mlu_op.h (hunk 9 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -4370,6 +4370,82 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
                    void *ious);
 
+// Group:NmsRotated
+/*!
+ * @brief Computes the index of nms with intersection-over-union
+ * (Jaccard index, IOU) of rotated bounding-boxes.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and
+ * queues in the nms_rotated operation. For detailed information, see
```
[the remark this answers is not in the document]
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 11 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 12 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 13 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: dets已修改为boxes

File: bangc-ops/mlu_op.h (hunk 14 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 15 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

**Review by GraceGuanH — APPROVED** [review]

File: bangc-ops/kernels/nms_rotated/nms_rotated_union1.mlu [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+               input_dx_ptr + input_offset + i * max_seg_iou_pad,
+               cpy_len * 1 * sizeof(IN_DT), boxes_load_dir);
+      // dy
+      __memcpy(box2 + half_offset + seg_len * 3,
+               input_dy_ptr + input_offset + i * max_seg_iou_pad,
+               cpy_len * 1 * sizeof(IN_DT), boxes_load_dir);
+      // angle
+      __memcpy(box2 + half_offset + seg_len * 4,
+               input_angle_ptr + input_offset + i * max_seg_iou_pad,
+               cpy_len * 1 * sizeof(IN_DT), boxes_load_dir);
```
Comment (working-bird) [inline bangc-ops/kernels/nms_rotated/nms_rotated_union1.mlu:220]: 198～220可以合并成一条memcpy嘛

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+| output_desc   | 输入数据，输出 index 的描述符  | 输入            | mluOpTensorDescriptor_t | /             | /        |
+| output        | 输出数据，输出 index 的数据    | 输出            | int32 \*                | [N]           | /        |
+| result_num    | 输出数据，输出 box 的个数      | 输出            | int32 \*                | /             | /        |
+
+### 1.4 算子限制
+
+| 限制类型     | 详细说明                                                                           |
+| -----------  | ---------------------------------------------------------------------------------- |
+| 输入限制     | 输入 `boxes`，`scores`的shape 必须满足要求: boxes[N, 5]或[N, 6] 和 scores:[N]      |
+| 输入限制     | 输入 `boxes`，`scores` 不支持输入 nan 或 inf。输入 `scores`不支持负数              |
```
Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:84]: "输入 `scores`不支持负数" 这块还是跟竞品对齐，竞品支持，我们方案上就应该支持，除非是指令导致，短时间难以绕过。

Comment (working-bird) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:84]: “输入 `boxes`，`scores` 不支持输入 nan 或 inf” -> 这个具体行为可以描述补充一下。另外这个要不要加到接口说明里。

File: bangc-ops/kernels/nms_rotated/nms_rotated_union1.mlu (hunk 17 above)
Comment (liuyuan1-v) [reply bangc-ops/kernels/nms_rotated/nms_rotated_union1.mlu]: 已合并

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 18 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已支持负数

Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已补充inf/nan行为

Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 接口中已添加

File: bangc-ops/kernels/nms_rotated/nms_utils.h [hunk 19]
```diff
@@ -0,0 +1,870 @@
+/*************************************************************************
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/nms_rotated/nms_utils.h:1]: licence 请更新一下

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+| 输出限制     | 输出 `output`的shape 必须满足: [N]                                                 |
+| 输出限制     | 输出 `result_num` 仅支持 int32\* 类型                                              |
+| 数据类型限制 | 输入 `boxes`，`scores` 数据类型保持一致且仅支持float。`output` 仅支持 int32        |
+| 原位限制     | 不支持原位                                                                         |
+| stride 限制  | 不支持 stride 机制                                                                 |
+| 广播限制     | 不支持广播                                                                         |
+
+限制说明：
+
+- 不支持`scores`中存在相同score的情况。竞品和mlu所选择框的index不一致。
```
Comment (DanieeelLiu) [inline docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md:95]: 竞品 --》 参考接口， 其他处也请检查下

File: docs/bangc-docs/user_guide/9_operators/index.rst [hunk 21]
```diff
@@ -656,3 +662,4 @@ Unique
 .. figure:: ../images/unique.png
 
 其中 ``x`` 表示输入数据，``y`` 表示输出数据。
+=======
```
Comment (WangYuan-cn) [inline docs/bangc-docs/user_guide/9_operators/index.rst:665]: 这

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -4524,6 +4524,81 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the output tensor, which indicates
+ * the index of each output box.
+ * @param[out] result_num
+ * Pointer to the MLU memory that stores the number of output boxes.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par Data Type
+ * - By the order of \b boxes - \b scores - \b output, the supported data types of
+ *    \b boxes, \b scores and \b output are as follows:
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:4567]: 这

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -4524,6 +4524,81 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ *   - float - float - int32
+ *
+ * @par Scale Limitation
+ * - The number of dimensions of \b boxes tensors must be 2.
+ * - The number of dimensions of \b scores and \b output tensors must be 1.
+ * - The highest dimension of \b boxes and \b scores must be equal.
+ * - The lowest dimension of \b boxes tensors must be 5 or 6.
+ *
+ * @note
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:4576]:
```suggestion
 * @par note
```

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -4524,6 +4524,81 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that needs to be used in the Nms operation.
+ * @param[in] output_desc
+ * The descriptor of the output tensor. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[out] output
+ * Pointer to the MLU memory that stores the output tensor, which indicates
+ * the index of each output box.
+ * @param[out] result_num
+ * Pointer to the MLU memory that stores the number of output boxes.
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:4561]:
```suggestion
 * Pointer to the MLU memory that stores the number of output boxes.
 *
```

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -4524,6 +4524,81 @@ mluOpBoxIouRotated(mluOpHandle_t handle,
...
+ * - The lowest dimension of \b boxes tensors must be 5 or 6.
+ *
+ * @note
+ * - The input \b boxes and \b scores with inf/nan are not supported currently. 
+ * 
+ * @par API Dependency
+ * - You need to call the ::mluOpGetNmsRotatedWorkspaceSize function to allocate extra
+ *   workspace for \b workspace.
+ *
+ * @par Requirements
```
Comment (WangYuan-cn) [inline bangc-ops/mlu_op.h:4583]: 这个删掉

**Review by WangYuan-cn — APPROVED** [review]

File: docs/bangc-docs/user_guide/9_operators/index.rst (hunk 21 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/user_guide/9_operators/index.rst]: 已删除

File: docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md (hunk 20 above)
Comment (liuyuan1-v) [reply docs/bangc-docs/design_docs/nms_rotated/nms_rotated.md]: 已修改

File: bangc-ops/kernels/nms_rotated/nms_utils.h (hunk 19 above)
Comment (liuyuan1-v) [reply bangc-ops/kernels/nms_rotated/nms_utils.h:1]: 已更新

File: bangc-ops/mlu_op.h (hunk 22 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已删除空格

File: bangc-ops/mlu_op.h (hunk 23 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

File: bangc-ops/mlu_op.h (hunk 24 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h:4595]: 已修改

File: bangc-ops/mlu_op.h (hunk 25 above)
Comment (liuyuan1-v) [reply bangc-ops/mlu_op.h]: 已修改

**Review by DanieeelLiu — APPROVED** [review]

**Review by cathyzhang222 — APPROVED** [review]

**Pull request merged**